TestClientSendAnomalies is unstable

### Which jobs are flaking?

unit test

### Which test(s) are flaking?

TestClientSendAnomalies

### Jenkins logs or GitHub Actions link

```log
[2022-03-22T03:15:01.116Z] --- FAIL: TestClientSendAnomalies (0.10s)
[2022-03-22T03:15:01.116Z]     client_test.go:345: 
[2022-03-22T03:15:01.116Z]         	Error Trace:	client_test.go:345
[2022-03-22T03:15:01.116Z]         	Error:      	An error is expected but got nil.
[2022-03-22T03:15:01.116Z]         	Test:       	TestClientSendAnomalies
```

https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_leak_test/detail/cdc_ghpr_leak_test/23015/pipeline

### Anything else we need to know

- Does this test exist for other branches as well?

- Has there been a high frequency of failure lately?

Another case

```log
[2022-03-24T06:29:27.318Z] --- FAIL: TestClientSendAnomalies (0.10s)
[2022-03-24T06:29:27.318Z]     client_test.go:345: 
[2022-03-24T06:29:27.318Z]         	Error Trace:	client_test.go:345
[2022-03-24T06:29:27.318Z]         	Error:      	An error is expected but got nil.
[2022-03-24T06:29:27.318Z]         	Test:       	TestClientSendAnomalies
```

https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_leak_test/detail/cdc_ghpr_leak_test/23229/pipeline

--- a/cdc/processor/agent_test.go
+++ b/cdc/processor/agent_test.go
@@ -66,8 +66,6 @@ type agentTestSuite struct {
 	blockSyncMu   sync.Mutex
 	blockSyncCond *sync.Cond
 	blockSync     bool
-
-	testMu sync.Mutex
 }
 
 func newAgentTestSuite(t *testing.T) *agentTestSuite {
@@ -498,30 +496,25 @@ func TestNoFinishOperationBeforeSyncIsReceived(t *testing.T) {
 	suite.tableExecutor.On("GetCheckpoint").
 		Return(model.Ts(1000), model.Ts(1000))
 
-	require.Never(t, func() bool {
-		suite.testMu.Lock()
-		defer suite.testMu.Unlock()
-
+	start := time.Now()
+	for time.Since(start) < 100*time.Millisecond {
 		err := agent.Tick(suite.cdcCtx)
 		require.NoError(t, err)
 
 		select {
 		case <-suite.ctx.Done():
-			return true
+			require.FailNow(t, "context is canceled")
 		case <-suite.dispatchResponseCh:
-			return true
+			require.FailNow(t, "Dispatch Response is received")
 		case <-suite.syncCh:
-			return true
+			require.FailNow(t, "Sync is received")
 		default:
-			return false
 		}
-	}, 100*time.Millisecond, 1*time.Millisecond)
+		time.Sleep(10 * time.Millisecond)
+	}
 	suite.UnblockSync()
 
 	require.Eventually(t, func() bool {
-		suite.testMu.Lock()
-		defer suite.testMu.Unlock()
-
 		select {
 		case <-suite.ctx.Done():
 			require.Fail(t, "context should not be canceled")

--- a/cdc/processor/agent_test.go
+++ b/cdc/processor/agent_test.go
@@ -66,6 +66,8 @@ type agentTestSuite struct {
 	blockSyncMu   sync.Mutex
 	blockSyncCond *sync.Cond
 	blockSync     bool
+
+	testMu sync.Mutex
 }
 
 func newAgentTestSuite(t *testing.T) *agentTestSuite {
@@ -497,6 +499,9 @@ func TestNoFinishOperationBeforeSyncIsReceived(t *testing.T) {
 		Return(model.Ts(1000), model.Ts(1000))
 
 	require.Never(t, func() bool {
+		suite.testMu.Lock()
+		defer suite.testMu.Unlock()
+
 		err := agent.Tick(suite.cdcCtx)
 		require.NoError(t, err)
 
@@ -514,6 +519,9 @@ func TestNoFinishOperationBeforeSyncIsReceived(t *testing.T) {
 	suite.UnblockSync()
 
 	require.Eventually(t, func() bool {
+		suite.testMu.Lock()
+		defer suite.testMu.Unlock()
+
 		select {
 		case <-suite.ctx.Done():
 			require.Fail(t, "context should not be canceled")

--- a/pkg/p2p/client_test.go
+++ b/pkg/p2p/client_test.go
@@ -341,9 +341,11 @@ func TestClientSendAnomalies(t *testing.T) {
 		time.Sleep(100 * time.Millisecond)
 		closeClient()
 	}()
-	_, err = client.SendMessage(ctx, "test-topic", &testMessage{Value: 1})
-	require.Error(t, err)
-	require.Regexp(t, ".*ErrPeerMessageClientClosed.*", err.Error())
+	_, _ = client.SendMessage(ctx, "test-topic", &testMessage{Value: 1})
+	// There is no need to check for error here, because when a client is closing,
+	// message loss is expected because sending the message is fully asynchronous.
+	// The client implementation is considered correct if `SendMessage` does not
+	// block infinitely.
 
 	wg.Wait()
 

